[WIP] Support swap Fn and CtlL keys (how it should be) for the x200, middleton

ypid opened the pull request:
For older Thinkpad models like the xx00 series the EC can not be configured to swap Fn and Ctrl. This setting would usually be exposed in the BIOS setup or in nvram when using coreboot. Enabled by default for models where this can not be configured.

Updates: #4 
Status: WIP, untested on real hardware. My x200 shipped with libreboot so I don’t have the factory rom which would come in handy to flash the EC :) Working on it.

hamishcoleman commented:
One reason why the generated x200 img file might not boot could be the fix-hdd-image-%i.patch files.

There are only two - one for each of two sizes of hdd image, so if the hdd is neither 65536 sectors nor 75776 sectors, there is no patch.

I've just added code to check for the lack of a matching disk image patch on the master branch.  If you do have a third disk image size, I probably should bite the bullet and write a tool to fix the partition table.

ypid commented:
> I probably should bite the bullet and write a tool to fix the partition table.

A generic solution is always preferred. Maybe one already exists? Have not checked yet.

ypid commented:
Lucky you, the old BIOS version used an 24117248 byte image (base for the patch by middleton) but the newer BIOS version uses a supported 38797312 byte image (same EC fw version). Forward porting the patch.

ypid commented:
Updated. I ported the patch to the latest EC firmware version to benefit from the battery charging fix from it. I confirmed that the ctrl fn swap works using all added patched images on a x200.

Unfortunately, just as I wanted to actually start using the laptop, I bricked it! I must have gotten the timing just right after the dos tool finished and rebooted into the bios (where the EC fw is transfered from SPI flash to EC flash as it seems) when I unplugged the USB DVD rom drive I tested with to confuse the BIOS and interrupt the EC flashing. But that was after I confirmed that `patched.x200.iso` works. I added a warning to the readme …

hamishcoleman commented on Makefile:
```diff
@@ -321,34 +352,38 @@ mec-tools/mec_encrypt: mec-tools/Makefile
 # $2 = FL2 filename
 define rule_fl2_patch
-    $(2): $(1) ; ./scripts/slice.insert $(1).slice $(1) $(2)
+    $(2): $(1) ./scripts/slice.insert ; ./scripts/slice.insert $(1).slice $(1) $(2)
```
Yes, technically this rule does depend on slice.insert, however changes to the slice.insert script (and similarly, the copyFL2 script below) should not change the contents of the resultant output.

(Conversely, there would not normally be any changes to these scripts, so no rebuilds would be triggered)

Can you expand a little on what your reasoning was?

ypid replied:
"should not change the contents of the resultant output." True, but we can elimiate this last bit of uncertanty also by simply rerunning the script if it changed.

I added this for "./scripts/slice.insert" as well when I was working on `copyFL2` which output did change. Does that work for you?

hamishcoleman replied:
Yes, that works for me.   I think I just dont like adding anything to those define lines - they are far too long.  But I should just consider this as more reasons to try and rework those lines into something less ugly.

hamishcoleman commented on autoexec.bat.command_com.template:
```diff
@@ -0,0 +1,24 @@
+echo.
+pause
+echo on
+command.com
```
Can you expand a little on what this is actually doing?  Have Lenovo built a custom command.com?  If so, what is interpreting this batch script?

ypid replied:
Not a batch script but some binary blob which was invoked exactly the same by the original autoexec.bat. I simply copied last line from the original file.

hamishcoleman replied:
Looking closer, it seems that /FLASH/COMMAND.COM is a compiled script which calls /FLASH/UPDTFLSH.EXE - another compiled script - which in turn calls FLASH/PHLASH16.EXE - and it looks like that one has some commandline options for specifying which files to flash.

I also wonder if the default flash scripts will allow you to re-flash a "same version" image - the xx30 tools certainly didnt, which is why zmatt worked out the dosflash commandline options.

Either way, to clean this up, it looks like some thought will be needed on how to make the autoexec bat generation better in the future.

ypid replied:
> I also wonder if the default flash scripts will allow you to re-flash a "same version" image

This seems not to be an issue here. The EC firmware could just be downgraded with the patched.x200.314.106 image/flasher. I can confirm that this worked. I also tried to increase the version numbers but this was not reflected in the BIOS message. To be continued.

> PHLASH16.EXE

This is worth a try, thanks!

hamishcoleman commented on autoexec.bat.template:
```diff
@@ -4,7 +4,8 @@ cd c:\flash
 echo.
-echo This will flash your firmware to support a classic 7-row keyboard
+echo This will flash your EC firmware to support a classic 7-row keyboard (default)
+echo or other modifications as configured at modification time.
```
The build system should probably keep track of which patches are enabled and applied and insert that here, to be 100% clear and correct

hamishcoleman commented on docs/CONFIG.txt:
```diff
@@ -1,9 +1,16 @@
+  For older Thinkpad models like the xx00 series the EC can not be configured
+  to swap Fn and Ctrl. This setting would usually be exposed in the BIOS setup
+  or in nvram when using coreboot.
+  Enabled by default for models where this can not be configured.
```
Possibly need to make it clear which patches are available for which types of laptop

hamishcoleman commented on scripts/checksum:
```diff
@@ -15,30 +17,30 @@ fi
+        rm "$FILE"
     fi
 
     exit 1
```
all these "fixes" patches should go into a separate patch and PR (similarly with some of the copyFL2 stuff below)

ypid replied:
True, I started with this one and moved it out of this PR and into #34. I will address your other comments in a few days :)

hamishcoleman commented on scripts/copyFL2:
```diff
@@ -9,54 +9,76 @@
+                exit 1
+            fi
+        done
+    fi
```
This is all way too complex.  I've been thinking about it and now think that the caller should simply provide an arg that is the pattern to match.  It could default to "FL2", but for the x200, it could be told to use "01B9x00.FL2" as the pattern, then the same "more than one match" logic could apply.

ypid replied:
It just works already. Your proposal would require more work for people adding such images with more than one FL2 file. Can‘t we just stay with this? I tried to validate all cases so that it should not break something.

hamishcoleman replied:
It doesnt require any additional work - there is already an assumption that you have the correct linux filename for the FL2 file (though, I see you have just used 'x' in that filename)

Also, I have already found ISO files from Lenovo where they have included multiple, unrelated FL2 files.

Finally, I'm trying to move towards a state where none of the inner names are hidden - and all are visible in the Makefile - which would then allow me to build an auto dependency maker and remove those hardcoded lists from the Makefile.  So adding this as a param actually makes work simpler in the future.

I have added the pattern matching feature described here to the master branch as I hope that might make it clearer.

ypid replied:
> Also, I have already found ISO files from Lenovo where they have included multiple

I guess there will always be this one case which breaks our assumption but requiring extra work for all normal cases where this can be handled automatically seems not right either.

> Finally, I'm trying to move towards a state where none of the inner names are hidden

Sounds good. They are not really hidden with this approach. Just that multiple FL2 files are updated when existing and started out equal. This can be made clear by using 'x' as you mentioned which I think is more transparent.

Ref: https://github.com/hamishcoleman/thinkpad-ec/commit/43506237f04e84557ed7b82f7822e11204411056#commitcomment-22825182

hamishcoleman commented:
Ouch! thats good to know about plugging things in during flashing, but sorry that you had to brick something to learn it.

There looked to be a Russian website that had EC flash-programmers, so you might find that there is a way to attach an in system programmer.

Do you have schematics for the x200?  I had a look at my x220 schematics and it looks like most of the JTAG pins were easily accessible (all but the CLK and RST lines are wired out to the KBDID$n pins on the x220)

ypid commented on Makefile:
```diff
@@ -141,6 +155,12 @@ patch_disable_keyboard:
+
+patch_disable_fn_ctrl_swap:
+	$(call patch_disable,008_fn_ctrl_swap.patch)
+
```
True. One the other hand, three patches are not that much either. If more get added then it should probably be optimized.

hamishcoleman replied:
agreed

ypid commented on Makefile:
```diff
@@ -228,6 +248,12 @@ $(DEPSDIR)/slice.insert.deps: Makefile
+	cp --reflink=auto $< $@
+	./mec-tools/mec_csum_outer -f $@
+	./mec-tools/mec_csum_outer -c $@
+
```
Excellent idea! This simplifies this case.

ypid commented on Makefile:
```diff
@@ -321,34 +352,38 @@ mec-tools/mec_encrypt: mec-tools/Makefile
+mec-tools/mec_csum_outer: mec-tools/Makefile
+	git submodule update
+	make -C mec-tools
+
```
Good catch, fixed.

ypid commented:
Yes, the manual mentions JTAG and also a on-board programming mode so it should be possible. And yes, I got lucky and have all the schematics :-) I already had a detailed look at the board with OpenBoardView. I will get back to this PR when I have tested it again. To speed this up, I ordered a second MB.

ypid commented:
Nice work on the master branch! I will need to rebase this PR later :)

I tested with a second MB and bricked it again. Seems I am becoming good at this :)

#### From BIOS 3.21/EC 1.06 -> BIOS 3.14/EC 1.06

* Did not reflash the EC

#### From BIOS 3.14/EC 1.06 -> BIOS 3.22/EC 1.06

* Warning mentions to either eject CD before reboot or after reboot finished, but not in between. Earlier versions did not which is why I bricked my first one :)
* Did not reflash the EC

#### From BIOS 3.22/EC 1.06 -> BIOS 3.14/EC 1.06 (increased EC version to 7XHT24WX)

* Did not reflash the EC

#### From BIOS 3.14/EC 1.06 -> BIOS 3.14/EC 1.09 7XHT26WW (manually increased EC version)

* Did reflash the EC
* Bricked the EC

```
@@ -32,9 +32,9 @@
 000001f0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 00000200  43 4f 52 50 2e 20 49 42  4d 20 32 30 30 31 20 41  |CORP. IBM 2001 A|
 00000210  4c 4c 20 52 49 47 48 54  53 20 52 45 53 45 52 56  |LL RIGHTS RESERV|
-00000220  45 44 00 00 00 00 00 00  10 60 b9 00 13 74 62 b4  |ED.......`...tb.|
+00000220  45 44 00 00 00 00 00 00  10 90 b9 00 13 74 62 b4  |ED...........tb.|
 00000230  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
-00000240  37 58 48 54 32 34 57 57  5f 5f 5f 5f 5f 5f 5f 5f  |7XHT24WW________|
+00000240  37 58 48 54 32 36 57 57  5f 5f 5f 5f 5f 5f 5f 5f  |7XHT26WW________|
 00000250  03 06 03 10 00 00 00 00  00 00 00 00 00 00 00 00  |................|
 00000260  00 00 00 00 28 43 29 20  43 6f 70 79 72 69 67 68  |....(C) Copyrigh|
 00000270  74 20 49 42 4d 20 43 6f  72 70 2e 20 32 30 30 31  |t IBM Corp. 2001|
```

hamishcoleman commented:
Now it sounds like you need to start practising hardware flashing :-(

It is good that the flash tools check the checksum - that at least means that we know we are not accidentally flashing bad data.  Did you do any tests with the phlash16.exe tool?

I know I made big changes to the master - effectively forcing a rebase, but what we have now is pretty close to what was in my head.  It should allow supporting multiple different major types of hardware in a straight-forward way.  There are still a bunch of things I can see to change, but I think it is time for me to go back to waiting until the feature is needed before ripping the infrastructure apart

ypid commented:
> Now it sounds like you need to start practising hardware flashing :-(

Sooner or later that will come in handy. Will probably require a bit of SMD soldering to change the mode of the EC and a second read of the data sheet.

> Did you do any tests with the phlash16.exe tool?

I gave it a quick try but got "Cannot flash when Memory Manager (e.g. HIMEM) are present".

ypid commented:
Just a little update: I still intend to finish this project, rebase the PR and work on getting it merged. But currently I don’t find much time. Maybe in a couple months. Two more working motherboards are waiting to be bricked (really cheap on ebay) and one day hopefully all of them can be un-bricked and put into service :)

hamishcoleman commented:
Thanks for the update - I've been busy and away myself, so have not had much time to spare either.

I might have some time before you do, so - if you dont have any objections - I might be able to rebase and start merging sooner than you (with appropriate warnings about bricking, it seems :-(

ypid commented:
Thanks! No objections from my side. About the bricking: It seems like the M$ Windows tool might be more resilient here. middleton provided patches for dozens of IBM/Lenovo models and users reported basically no issues. Anyway, it has to work without Windows :)

herbsmn commented:
I have an x200 with libreboot on it and would love to liberate the EC. It sounds like work is still being done to make this process stable though. Is that correct?

ypid commented:
Not right now unfortunately as far as I know. I linked to a few other places. Also swiftgeek is interested in this and might be able to help you. Feel free to pick up where we left off.

herbsmn commented:
So does this support any libreboot supported laptops? I also asked this question over here: https://github.com/lynxis/frenc/issues/1

herbsmn commented:
@ypid: What model of libreboot laptop were you using with @lynxis' liberated h8s EC? I assume that this thinkwiki page doesn't enumerate all of the Thinkpad models that have h8s: https://www.thinkwiki.org/wiki/Renesas_H8S/2161BV

ypid commented:
I can not help you much here. I only flashed the stock/factory EC firmware (slightly patched) using the updated mechanism which is supposed to be used by end users (done as part of system firmware update). I have not played with the EC firmware rewrite but I am aware of it and would appreciate a free EC firmware.

herbsmn commented:
@ypid Thanks for the response. I misread your last comment. I too would appreciate a free EC firmware.
